parse/procedures/visitors/typing/typing.go

```diff
@@ -579,49 +628,68 @@ func (t *typingVisitor) analyzeProcedureCall(p0 parser.ICallExpression) *types.P
 }
 
 func (t *typingVisitor) VisitStatementBreak(p0 *parser.StatementBreak) any {
+	if t.loopTarget == "" {
+		t.errs.NodeErr(p0.GetNode(), types.ParseErrorTypeSemantic, types.ErrBreakUsedOutsideOfLoop)
+	}
 	return nil
 }
 
 func (t *typingVisitor) VisitStatementForLoop(p0 *parser.StatementForLoop) any {
+	// set the loop target so children know we are in a loop
 	t.loopTarget = p0.Variable
 
 	switch target := p0.Target.(type) {
 	case *parser.LoopTargetVariable:
-		r := target.Accept(t).(*types.DataType)
+		r := target.Accept(t).(*coreTypes.DataType)
 
 		// we will not declare these as anonymous, since this is simply
 		// a field in an array of a known type
 		_, ok := t.declarations[p0.Variable]
 		if ok {
-			panic("variable already declared")
+			t.errs.NodeErr(p0.GetNode(), types.ParseErrorTypeSemantic, types.ErrVariableAlreadyDeclared)
+			return nil
 		}
 
 		t.declarations[p0.Variable] = r
+
+		// check that we are iterating over an array
+		dataType, ok := t.declarations[target.Variable.Name]
+		if !ok {
+			t.errs.NodeErr(p0.GetNode(), types.ParseErrorTypeSemantic,
```

Comment:
I see this is not consistent with VisitExpressionVariable, you returned type error there